Docs/http api (#355)
stoplightio/prism: packages/http/README.md

# Prism HTTP

This package provides a http client featuring the ability to:

- mock requests instead of hitting the server
- validate the request and the response according to an openapi spec

# Basic Usages

> Note: examples in this document use the following spec

```yaml
---
openapi: 3.0.2
paths:
  /todos:
    get:
      parameters:
        - name: title
          in: query
          style: form
          schema:
            enum:
              - eat
              - drink
      responses:
        200:
          description: Get Todo Items
```

> ## I want to use Prism to mock responses

Reviewer: Could we add a quick sentance on why? 

Talk about how it can let you mock without needing to hit an actual server instance, which is great for browser-based mocking, etc.

[Try it!](https://repl.it/@ChrisMiaskowski/prism-http-client-basic-mocking);

```javascript
const Prism = require('@stoplight/prism-http');
const path = require('path');

// Create Prism instance and configure it as a mocker generating static examples
const config = { mock: { dynamic: false } };
const prism = Prism.createInstance(config);

// Load openapi spec file
const specPath = path.resolve(process.cwd(), 'basic.oas3.yaml');
prism
  .load({ path: specPath })
  .then(() => {
    // Make a "GET /todos" request
    return prism.process({
      method: 'get',
      url: {
        path: '/todos',
      },
    });
  })
  .then(prismResponse => {
    console.log(prismResponse.output);
  });
```

Output

```bash
{ statusCode: 200,
  headers: { 'Content-type': 'text/plain' },

>   body: undefined }

Reviewer: Do we really want an undefined body as the primary example for a response mocking tool? If feels like we are missing an opportunity to actually show off what prism does.

Reviewer 2: Yeah good point :P

```

> ## I want Prism to make a http request a server

Reviewer: This might not be clear enough for people without context. We want to explain that Prism HTTP Client is a generic HTTP client first, and it also happens to have mock mode enabled. 

Imagine we are releasing a axios competitor, this README https://www.npmjs.com/package/axios should be an inspiration. 

One of the features of our HTTP client is that it can shortcut and return mocks instead. 

Instead of talking to localhost:4010 which might confuse people and make them think they are talking to Prism HTTP Server, why don't we make calls to https://pokeapi.co/ or api.example.org or literally anything other than localhost.

Reviewer 2: > Instead of talking to localhost:4010 which might confuse people and make them think they are talking to Prism HTTP Server, why don't we make calls to https://pokeapi.co/ or api.example.org or literally anything other than localhost.

This is exactly what I did :) I tested it using the prism CLI

We could use api.example.org or pokeapi.co buuuuut api.example.org does not return stuff... pokeapi.co doesn't do anything relevant either.

Instead, if someone runs the Prism CLI with the spec provided at the top of the documentation they will be able to actually test this stuff.

Reviewer: Ok, I like the idea that this should definitely actually run. Pokeapi is an API with JSON data so if you take a second look you'll see you can construcvt URLs.

https://pokeapi.co/api/v2/pokemon/ditto/

JSON! 

I respect that using the CLI is a good way to have a server running locally that we know they have, but can you see how it is confusing for the user? Using prism to talk to prism is meta and conflates two different components and their use.

Reviewer 2: Gotcha! Let's GO with Pokemons! ;)

Reviewer 2: Actually, I picked cats :P

# Prism HTTP

> This package provides HTTP APIs for forwarding and mocking requests.

Reviewer: The HTTP client package is meant to be a generic HTTP client which has the ability to also mock responses. We should be advertising it like that, and we want to avoid talking about forwarding because that’s an internal implementation detail of a proxy we no longer use.

> # Http Client

Reviewer 3: "HTTP" over "Http"

This package provides a http client featuring the ability to:

> - mock responses based of OAS spec

Reviewer 3: "OpenAPI" over "OAS"

- make requests to a http server
- validate the request and/or the response according to the provided OAS

The goal of this document is to provide you with some basic code examples to get you started and to cover some of the advanced scenarios.

# Installation

> `yarn add @stoplight/prism-http`

Reviewer: ```suggestion
```
yarn add @stoplight/prism-http
```
```

# Basic Usages

> Note: examples in this document use the following spec

```yaml
---
openapi: 3.0.2
paths:
  /todos:
    get:
      parameters:
        - name: title
          in: query
          style: form
          schema:
            enum:
              - eat
              - drink
      responses:
        200:
          description: Get Todo Items
```

## I want to mock responses of operations defined in an OAS file

> [Try it out!](https://repl.it/@ChrisMiaskowski/prism-http-client-basic-mocking)

Reviewer 3: This is cool!

Reviewer 2: Hey, I have built similar under personal account for Spectral. let's get a SL one and I will replace these at that point.

```javascript
const Prism = require('@stoplight/prism-http');
const path = require('path');

// Create Prism instance and configure it as a mocker generating static examples

> const config = { mock: { dynamic: false } };

Reviewer 4: It does not.

Mock is either `disabled` or configured accordingly.

const prism = Prism.createInstance(config);

// Load openapi spec file
const specPath = path.resolve(process.cwd(), 'basic.oas3.yaml');
prism
  .load({ path: specPath })
  .then(() => {
    // Make a "GET /todos" request
    return prism.process({
      method: 'get',
      url: {
        path: '/todos',
      },
    });
  })
  .then(prismResponse => {
    console.log(prismResponse.output);
  });
```

Output

```bash
{ statusCode: 200,
  headers: { 'Content-type': 'text/plain' },

>   body: undefined }

Reviewer 4: ```suggestion
  body: undefined 
}
```

```

> ## I want to make a http request a server

Reviewer 4: ```suggestion
## Proxying requests to an upstream server
```

> In this use case we assume we have a server running at `http://localhost:4010`

Reviewer 4: This is a little bit confusing; this is usually the address Prism listens to. Maybe let's assume the server is running on 3030 or 8080?

Reviewer: @XVincentX i was suggesting a slightly different direction, to make it even clearer, we could use not localhost.

> Instead of talking to localhost:4010 which might confuse people and make them think they are talking to Prism HTTP Server, why don't we make calls to https://pokeapi.co/ or api.example.org or literally anything other than localhost.

that is able to handle `GET /todos` request.

We don't want to mock a reqeust, we simply want to make the request, hit the actual server and get the response back.

```javascript
const Prism = require('@stoplight/prism-http');

// Create Prism instance and configure it to make http requests (mock: false)
const config = { mock: false };
const prism = Prism.createInstance(config);
prism
  .process({
    method: 'get',
    url: {
      path: '/todos',
      baseUrl: 'http://localhost:4010',
    },
  })
  .then(prismResponse => {
    console.log(prismResponse.output);
  });
```

In response you'll get whatever the server responds with.

> ## I want all my requests to talk to a server unless I override it per a reqeust

Reviewer 4: ```suggestion
## Overriding forwarding to mock for a single request
```

prism
  .load({ path: specPath })
  .then(() => {
    // Make a "GET /todos" request
    return prism.process(
      {
        method: 'get',
        url: {
          path: '/todos',
          baseUrl: 'http://localhost:4010',
        },
      },
      {
        // We can override the default behaviour per request.
        mock: {
          dynamic: true,
        },
      }
    );
  })
  .then(prismResponse => {
    console.log(prismResponse.output);
  });
```

# Advanced Topics

## Creating Prism Instance

In order to create and instance of Http Client (later referred to as `prism` for simplicity) use the `createInstance` function, like so:

```javascript
const Prism = require('@stoplight/prism-http');
const config = { mock: false };
const prism = Prism.createInstance(config /*, components */);
```

There are two (both optional) arguments you can supply `createInstance` with:

- config (of `IHttpConfig` type)
- components (of `TPrismHttpComponents` type)

We will cover the `config` argument in next section and we'll leave `components` for some other time (overriding default `components` is the _ultimate advanced stuff_).

> ### Config Property

Reviewer 4: ```suggestion
### Config Object
```

Prism's config object (`IHttpConfig`) allows you to manipulate Prism's behaviour in many ways. For instance:

- turn validation on and off
- turn mocking on and off
- influence Prism's mocked response generation strategy

> The actual interface looks like this (but rather than explain each property I will give you some examples)

Reviewer 2: I did that intentionally. I felt a reference would be less useful. I think most developers would rather see a bunch of examples rather than description of types. Often those descriptions are not telling much.

Plus I think if we were to simply document types... We can automatically generate such documentation with jsdoc or whatever.

> The actual interface looks like this (but rather than explain each property I will give you some examples)

Reviewer 4: Automatically generated docs: nobody will never look at it. Do you remember our fantastic TSDocs for Graphite and other projects? Did somebody ever look at them?

We eliminated them at the end.

I'm ok with the format; we're going to need a complete reference probably at some point though.

```ts
export interface IHttpConfig extends IPrismConfig {
  mock: false | IHttpOperationConfig;

>   validate?: {

Reviewer 4: This is stuff that non TypeScript people might struggle to understand.

Reviewer 2: Any suggestions? The reason I copied this he is I didn't want to introduce yet another format for documenting our interfaces.

Reviewer 4: Eh, not right now unfortunately. Maybe revisiting the API could help us to make it better, so I guess we'll keep it as it is right now.

          body?: boolean;
        };
  };
}
```

#### Config Examples

**Do not mock and do not validate requests**

```javascript
const config = {
  mock: false,
  validate: {
    request: false,
    // response is set to true by default
  },
};
```

With the above configuration the http client will proxy your requests to a server and will validate the response with the OAS you loaded. However, it will not validate the input (e.g. will not check whether the provided query param is valid).

**Validate only request's body and response's headers**

```javascript
const config = {
  validate: {
    request: {
      body: true,
    },
    response: {
      headers: true,
    },
  },
};
```

**When mocking a response generate static response**

```javascript
const config = {
  mock: {
    dynamic: false,
  },
};
```

Generating static responses means that if an OAS operation's response has any examples defined the payload we construct will use those examples.

This contrasts "dynamic responses" which means generating responses from a json schema.

**Return named example of a specified key**

```javascript
const config = {
  mock: {
    example: 'key',
  },
};
```

> In OAS3 one can construct such response that will have multiple examples mapped by keys.

Reviewer 4: ```suggestion
In OpenAPI Specification 3 one can construct such response that will have multiple examples mapped by keys.
```

This configuration allows you to be very specific which example you want to choose.

**Return 403 response of specified mime type**

```javascript
const config = {
  mock: {
    code: 403,
    mimeType: 'application/xml',
  },
};
```

This will enforce a 403 response (given that such response is defined in your OAS file).

## Loading specs

When you load a spec you "teach" prism how your API looks like.
You can run prism without loading a spec, but features such as validation will be disabled.

```javascript

> const loaderConfig = { path: '/absolute/file/path.oas3.yaml' };

Reviewer 2: Sure, lots of this stuff will eventually change :) Let's update once it does

The goal of this document is to provide you with some basic code examples to get you started and to cover some of the advanced scenarios.

# Table of Contents

- [Installation](#installation)
- [Basic Usages](#basic-usages)
- [Advanced Topics](#advanced-topics)
  - [Creating Prism Instance](#creating-prism-instance)
    - [Config Property](#config-property)
      - [Config Examples](#config-examples)
  - [Loading specs](#loading-specs)
  - [Making requests](#making-requests)
    - [Server Validation](#server-validation)
  - [Understanding response](#understanding-response)
  - [Gotchas](#gotchas)

# Installation

```bash
yarn add @stoplight/prism-http
```

# Basic Usages

> Note: the examples in this document use the following spec

```yaml
openapi: 3.0.2
paths:
  /todos:
    get:
      parameters:
        - name: title
          in: query
          style: form
          schema:
            enum:
              - eat
              - drink
      responses:
        200:
          description: Get Todo Items
          content:
            'text/plain':

>               example: Write lit OpenAPI specs!

Reviewer: ```suggestion
              example: Write dope OpenAPI specs!
```

@chris-miaskowski you missed one

```

## Mock All Responses

[Try it out!](https://repl.it/@ChrisMiaskowski/prism-http-client-basic-mocking)

```javascript
const Prism = require('@stoplight/prism-http');
const path = require('path');

// Create Prism instance and configure it as a mocker generating static examples
const prism = Prism.createInstance({
  mock: { dynamic: false },
  validate: { request: { query: true } },
});

// Load openapi spec file
const specPath = path.resolve(process.cwd(), 'basic.oas3.yaml');
prism
  .load({ path: specPath })
  .then(() => {
    // Make a "GET /todos" request
    return prism.process({
      method: 'get',
      url: {
        path: '/todos',
      },
      headers: {
        Accept: 'text/plain',
      },
    });
  })
  .catch(e => console.error(e))
  .then(prismResponse => {
    console.log(prismResponse.output);
  });
```

Output

```bash
{
  statusCode: 200,
  headers: { 'Content-type': 'text/plain' },

>   body: 'Write lit OpenAPI specs!'

Reviewer: ```suggestion
  body: 'Write literally any other word than lit OpenAPI specs!'
```

**Return 403 response of specified mime type**

```javascript
const config = {
  mock: {
    code: 403,
    mimeType: 'application/xml',
  },
};
```

This will enforce a 403 response (given that such response is defined in your OpenAPI file).

## Loading specs

When you load a spec you "teach" prism how your API looks like.
You can run prism without loading a spec, but features such as validation will be disabled.

```javascript
const loaderConfig = { path: '/absolute/file/path.oas3.yaml' };
const promise = prism.load(loaderConfig);
```

## Making requests

To make a basic request you need to do the following

```javascript
const request = {
  method: 'get',
  url: {
    path: '/path', // must be prefixed with slash
  },
};
const promise = prism.process(request);
```

The request object has the following interface

```ts
export interface IHttpRequest {
  method: IHttpMethod;
  url: IHttpUrl;
  headers?: IHttpNameValue;
  body?: any;
}
```

All of this is pretty standard except the `url.baseUrl` which we will describe in more details.

### Server Validation

Consider a request

```javascript
prism.process({
  method: 'get',
  url: {
    path: '/path', // must be prefixed with slash

>     baseUrl: 'http://localhost:4010',

Reviewer: Are we keeping localhost:4010 here for a reason? We changed the other instances to a real server for clarity, is this different ?

Reviewer 2: we are not, I'm dumb :P
